### ISSUE TITLE
Realm file decryption failed

### ISSUE THREAD

**Issue by Goner** [realm/realm-cocoa#4257]
After I encrypting the realm file, sometimes, when I  run app again, I can get key and an error "Realm file decryption failed". But I can open realm file without key by realm browser.

when I find it seems that I fail to encrypt the file, then I set config's encryptkey nil, get an error "Unsupported Realm file format version" 

now I cant know whether I Encryption failed or I Encrypt in wrong way
## Code Sample encryptionKey

```
+ (NSData *)getKey {
    // Identifier for our keychain entry - should be unique for your application
    static const uint8_t kKeychainIdentifier[] = "jl.com.app";
    NSData *tag = [[NSData alloc] initWithBytesNoCopy:(void *)kKeychainIdentifier
                                               length:sizeof(kKeychainIdentifier)
                                         freeWhenDone:NO];

    // First check in the keychain for an existing key
    NSDictionary *query = @{(__bridge id)kSecClass: (__bridge id)kSecClassKey,
                            (__bridge id)kSecAttrApplicationTag: tag,
                            (__bridge id)kSecAttrKeySizeInBits: @512,
                            (__bridge id)kSecReturnData: @YES};

    CFTypeRef dataRef = NULL;
    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query, &dataRef);
    if (status == errSecSuccess) {
        return (__bridge NSData *)dataRef;
    }

    // No pre-existing key from this application, so generate a new one
    uint8_t buffer[64];
    SecRandomCopyBytes(kSecRandomDefault, 64, buffer);
    NSData *keyData = [[NSData alloc] initWithBytes:buffer length:sizeof(buffer)];

    // Store the key in the keychain
    query = @{(__bridge id)kSecClass: (__bridge id)kSecClassKey,
              (__bridge id)kSecAttrApplicationTag: tag,
              (__bridge id)kSecAttrKeySizeInBits: @512,
              (__bridge id)kSecValueData: keyData};

    status = SecItemAdd((__bridge CFDictionaryRef)query, NULL);
    NSAssert(status == errSecSuccess, @"Failed to insert new key in the keychain");

    return keyData;
}
```
## Version of Realm and Tooling

Realm version: 1.0.2

Xcode version: xcode8

iOS/OSX version: ios 10

**Comment by TimOliver** [comment]
Hi @Goner!

Hmm, if you properly encrypted the Realm, you shouldn't be able to open it in the Realm Browser without providing the same key. It sounds like you may have created an unencrypted Realm and are trying to encrypt it after the fact.

In order to create an encrypted Realm, you need to ensure you've set the `encryptionKey` property in its `Configuration` before the first time you create a `Realm()` instance. 

```
RLMConfiguration *config = [RLMRealmConfiguration defaultConfiguration];
config.encryptionKey = [[self class] getKey];
[RLMRealmConfiguration setDefaultConfiguration:config];

RLMRealm *realm = [RLMRealm defaultRealm];
```

If you don't specify the key before creating `RLMRealm` for the firs time, the file will be unencrypted. It's not possible to convert an unencrypted Realm into an encrypted one; nor is it possible to change the encryption key once it's set. To do either, you'll need to use [`- [RLMRealm writeCopyToURL:encryptionKey:error:]`](https://realm.io/docs/objc/2.0.2/api/Classes/RLMRealm.html#/c:objc(cs)RLMRealm(im)writeCopyToURL:encryptionKey:error:) to write a whole new Realm file with those settings.

I hope that helped! Let me know if you have any more questions.

**Comment by Goner** [comment]
thanks for your help,
 I'm sure that I specify the key before creating a new RLMRealm for the first time. But the key sometimes when I run app again and get the encryptionKey from my keychain,  cant open the realm, error with "Realm file decryption failed". This situation is hard to reappear. Only once I catch it, then find I can open it without key. BUT,I can get the key in my keychain.
I check my code about encrypting, I think its work. 


CODE:
+ (void)setDefaultRealmForUser:(NSString *)username {
    RLMRealmConfiguration *config = [RLMRealmConfiguration defaultConfiguration];
    
    config.encryptionKey = [self getKey];
    
    
    config.fileURL = [[[config.fileURL URLByDeletingLastPathComponent]
                       URLByAppendingPathComponent:username]
                      URLByAppendingPathExtension:@"realm"];
    
    
    [RLMRealmConfiguration setDefaultConfiguration:config];
}

this is crash log：
lichade1995.realm: Unable to open a realm at path '/var/mobile/Containers/Data/Application/85C75293-3A36-4F68-BE19-BA25118B7F69/Documents/lichade1995.realm': Realm file decryption failed.

This problem is Seriously affected my app....

**Comment by Goner** [comment]
Actually, I find it always crash at I open the default realm while I create a new realm(new url) with encryptionKey. but in fact they are truely two different realm. Does it matter？

**Comment by Goner** [comment]
I think that the encryptionKey for my encrypted realm is right. I can open encrypted realm after the crash(app restrat). If the encryptionKey is wrong, should always crash. Now, I dont know what's the problem about the crash.

**Comment by Goner** [comment]
If the default realm has been opened, get some information from it.At the same time, I  generate a new encrypted realm.There will be a problem about  "Realm file decryption failed,Unable to open a realm at path "?

**Comment by TimOliver** [comment]
If they are two different Realms, then yes, it would matter. 

Hmm, if you've already set a default Realm, and already opened it, it's probably not a good idea to call `[RLMRealmConfiguration setDefaultConfiguration:]` again to rotate the default Realm to a different one. Once an `RLMRealm` instance is created off a configuration, copies of it are kept in memory after the fact to improve future access efficiency.

If that's the case, instead of relying on the default Realm mechanism, it may be easier to manually control access to the user Realm using `[RLMRealm realmWithConfiguration:]`. All you'd need is a static class to generate the appropriate `RLMRealmConfiguration` object for the user, and then create the `RLMRealm` instance each time you need it.

**Comment by Goner** [comment]
Is there a way to check the encryptionKey? Do Not Crash Way

**Comment by jpsim** [comment]
@Goner I don't understand your last request. `encryptionKey` is a property of `RLMRealmConfiguration`, and reading its value is just reading an inert `NSData` property. That is a "Do Not Crash Way" to access its value.

If you're asking if there's a way to derive the encryption key used to encrypt a Realm, the answer is obviously no because that would defy the purpose of encrypting it.

**Comment by jpsim** [comment]
Closing due to lack of updates.

**Comment by Goner** [comment]
Fatal Exception: RLMException
/var/mobile/Containers/Data/Application/55C02158-FF89-4E93-94D0-A2BC9180960C/Documents/q2013394194.realm: Unable to open a realm at path '/var/mobile/Containers/Data/Application/55C02158-FF89-4E93-94D0-A2BC9180960C/Documents/q2013394194.realm': Realm file decryption failed.

I'm sure that the key is right. Because I save the key in NSUserDefaults after I open the realm.But in some users' crash log, 'Realm file decryption failed' always happened. And I'm sure the 'q2013394194.realm' is the ecrypted realm.

This problem has troubled us for a long time...

**Comment by jpsim** [comment]
@Goner please file a new issue with steps to reproduce so we can investigate this for you.
